File: antidescuadre/src/logica/configuracion.ts
import { db, obtenerAjustes, guardarAjustes } from '../db/db'
import type { ConfigExportada } from '../db/tipos'

// Exporta SOLO configuración (catálogo, categorías, variantes, precios,
// alias de mesas, ajustes). Nunca datos operativos de turno (regla 15).
export async function exportarConfiguracion(): Promise<ConfigExportada> {
  const [ajustes, categorias, productos, mesas] = await Promise.all([
    obtenerAjustes(),
    db.categorias.toArray(),
    db.productos.toArray(),
    db.mesas.orderBy('orden').toArray(),
  ])
  return {
    app: 'antidescuadre',
    version: 1,
    exportadoEn: new Date().toISOString(),
    nombreNegocio: ajustes.nombreNegocio,
    alertaMinutos: ajustes.alertaMinutos,
    categorias: categorias.map(c => ({ id: c.id!, nombre: c.nombre, padreId: c.padreId })),
    productos: productos.map(p => ({
      nombre: p.nombre, precio: p.precio, categoriaId: p.categoriaId,
      activo: p.activo, grupos: p.grupos,
    })),
    mesas: mesas.map(m => ({ alias: m.alias, orden: m.orden })),
  }
}

export function validarConfiguracion(dato: unknown): ConfigExportada | null {
  if (typeof dato !== 'object' || dato === null) return null
  const c = dato as Record<string, unknown>
  if (c.app !== 'antidescuadre' || c.version !== 1) return null
  if (!Array.isArray(c.categorias) || !Array.isArray(c.productos) || !Array.isArray(c.mesas)) return null
  return c as unknown as ConfigExportada
}

// Reemplaza la configuración actual por la importada. No toca datos operativos
// (turnos, cuentas, ítems, pagos, comprobantes).
export async function importarConfiguracion(config: ConfigExportada): Promise<void> {
  await db.transaction('rw', [db.categorias, db.productos, db.mesas, db.ajustes], async () => {
    await db.categorias.clear()
    await db.productos.clear()
    await db.mesas.clear()

    // Reinsertar categorías conservando la jerarquía con ids nuevos
    const mapaIds = new Map<number, number>()
    const pendientes = [...config.categorias]
    let vueltas = 0
    while (pendientes.length > 0 && vueltas < 100) {
      vueltas++
      for (let i = pendientes.length - 1; i >= 0; i--) {
        const cat = pendientes[i]
        if (cat.padreId === null || mapaIds.has(cat.padreId)) {
          const nuevoId = await db.categorias.add({
            nombre: cat.nombre,
            padreId: cat.padreId === null ? null : mapaIds.get(cat.padreId)!,
          })
          mapaIds.set(cat.id, nuevoId)
          pendientes.splice(i, 1)
        }
      }
    }

    for (const p of config.productos) {
      await db.productos.add({
        nombre: p.nombre,
        precio: p.precio,
        categoriaId: p.categoriaId != null ? (mapaIds.get(p.categoriaId) ?? null) : null,
        activo: p.activo !== false,
        grupos: Array.isArray(p.grupos) ? p.grupos : [],
      })
    }

    await db.mesas.bulkAdd(config.mesas.map((m, i) => ({ alias: m.alias, orden: m.orden ?? i })))
  })
  await guardarAjustes({
    nombreNegocio: config.nombreNegocio ?? 'Mi bar',
    alertaMinutos: config.alertaMinutos ?? 10,
  })
}

export function descargarArchivo(nombre: string, contenido: string): void {
  const blob = new Blob([contenido], { type: 'application/json' })
  const url = URL.createObjectURL(blob)
  const a = document.createElement('a')
  a.href = url
  a.download = nombre
  a.click()
  setTimeout(() => URL.revokeObjectURL(url), 5000)
}

File: app/lib/logica/configuracion.dart
// Exportar/importar configuración (sección 3.11, regla 15): SOLO catálogo,
// categorías, variantes, precios, alias de mesas y ajustes. Nunca datos
// operativos de un turno (mesas activas, cuentas, ventas, comprobantes).

import 'dart:convert';

import 'package:drift/drift.dart';

import '../datos/base.dart';

Future<Map<String, dynamic>> exportarConfiguracion(BaseDatos db) async {
  final ajustes = await db.select(db.ajustes).getSingleOrNull();
  final categorias = await db.select(db.categorias).get();
  final productos = await db.select(db.productos).get();
  final mesas = await (db.select(db.mesas)
        ..orderBy([(m) => OrderingTerm.asc(m.orden)]))
      .get();
  return {
    'app': 'antidescuadre',
    'version': 2,
    'exportadoEn': DateTime.now().toIso8601String(),
    'nombreNegocio': ajustes?.nombreNegocio ?? 'Mi bar',
    'alertaMinutos': ajustes?.alertaMinutos ?? 10,
    'simboloMoneda': ajustes?.simboloMoneda ?? '\$',
    'formato24h': ajustes?.formato24h ?? false,
    'recordatorioBackupDias': ajustes?.recordatorioBackupDias ?? 0,
    'vibracionActiva': ajustes?.vibracionActiva ?? true,
    'categorias': [
      for (final c in categorias) {'id': c.id, 'nombre': c.nombre, 'padreId': c.padreId},
    ],
    'productos': [
      for (final p in productos)
        {
          'nombre': p.nombre,
          'precio': p.precio,
          'categoriaId': p.categoriaId,
          'activo': p.activo,
          'grupos': jsonDecode(p.gruposJson),
        },
    ],
    'mesas': [
      for (final m in mesas) {'alias': m.alias, 'orden': m.orden},
    ],
  };
}

Map<String, dynamic>? validarConfiguracion(Object? dato) {
  if (dato is! Map<String, dynamic>) return null;
  // Acepta versión 1 (archivos viejos) y 2; los campos nuevos que falten
  // caen a sus valores por defecto al importar.
  if (dato['app'] != 'antidescuadre' || (dato['version'] != 1 && dato['version'] != 2)) {
    return null;
  }
  if (dato['categorias'] is! List || dato['productos'] is! List || dato['mesas'] is! List) {
    return null;
  }
  return dato;
}

// Reemplaza la configuración actual por la importada, reconstruyendo la
// jerarquía de categorías con ids nuevos. No toca datos operativos.
Future<void> importarConfiguracion(BaseDatos db, Map<String, dynamic> config) async {
  await db.transaction(() async {
    await db.delete(db.categorias).go();
    await db.delete(db.productos).go();
    await db.delete(db.mesas).go();

    final mapaIds = <int, int>{};
    final pendientes = List<Map<String, dynamic>>.from(
      (config['categorias'] as List).map((c) => Map<String, dynamic>.from(c as Map)),
    );
    var vueltas = 0;
    while (pendientes.isNotEmpty && vueltas < 100) {
      vueltas++;
      for (var i = pendientes.length - 1; i >= 0; i--) {
        final cat = pendientes[i];
        final padre = cat['padreId'] as int?;
        if (padre == null || mapaIds.containsKey(padre)) {
          final nuevoId = await db.into(db.categorias).insert(CategoriasCompanion.insert(
                nombre: cat['nombre'] as String,
                padreId: Value(padre == null ? null : mapaIds[padre]),
              ));
          mapaIds[cat['id'] as int] = nuevoId;
          pendientes.removeAt(i);
        }
      }
    }

    for (final p in config['productos'] as List) {
      final prod = Map<String, dynamic>.from(p as Map);
      final grupos = prod['grupos'];
      await db.into(db.productos).insert(ProductosCompanion.insert(
            nombre: prod['nombre'] as String,
            precio: (prod['precio'] as num).toDouble(),
            categoriaId: Value(
              prod['categoriaId'] == null ? null : mapaIds[prod['categoriaId'] as int],
            ),
            activo: Value(prod['activo'] != false),
            gruposJson: Value(grupos is List ? jsonEncode(grupos) : '[]'),
          ));
    }

    var i = 0;
    for (final m in config['mesas'] as List) {
      final mesa = Map<String, dynamic>.from(m as Map);
      await db.into(db.mesas).insert(MesasCompanion.insert(
            alias: mesa['alias'] as String,
            orden: Value(mesa['orden'] as int? ?? i),
          ));
      i++;
    }
  });
  await db.guardarAjustes(
    nombreNegocio: config['nombreNegocio'] as String? ?? 'Mi bar',
    alertaMinutos: config['alertaMinutos'] as int? ?? 10,
    simboloMoneda: config['simboloMoneda'] as String? ?? '\$',
    formato24h: config['formato24h'] as bool? ?? false,
    recordatorioBackupDias: config['recordatorioBackupDias'] as int? ?? 0,
    vibracionActiva: config['vibracionActiva'] as bool? ?? true,
  );
}
